controllers/components/datasciencepipelines/datasciencepipelines_controller_actions.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package datasciencepipelines
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	componentsv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/components/v1"
+	"github.com/opendatahub-io/opendatahub-operator/v2/controllers/status"
+	odhtypes "github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+	odhdeploy "github.com/opendatahub-io/opendatahub-operator/v2/pkg/deploy"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/metadata/labels"
+)
+
+func initialize(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
+	rr.Manifests = append(rr.Manifests, odhtypes.ManifestInfo{
+		Path:       DefaultPath,
+		ContextDir: "",
+		SourcePath: "",
+	})
+	if err := odhdeploy.ApplyParams(DefaultPath, nil, map[string]string{"namespace": rr.DSCI.Spec.ApplicationsNamespace}); err != nil {
+		return fmt.Errorf("failed to update params.env from %s : %w", rr.Manifests[0], err)
+	}
+	return nil
+}
+
+func devFlags(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
+	dsp, ok := rr.Instance.(*componentsv1.DataSciencePipelines)
+	if !ok {
+		return fmt.Errorf("resource instance %v is not a componentsv1.DataSciencePipelines)", rr.Instance)
+	}
+
+	if dsp.Spec.DevFlags == nil {
+		return nil
+	}
+
+	// Implement devflags support logic
+	// If dev flags are set, update default manifests path
+	if len(dsp.Spec.DevFlags.Manifests) != 0 {
+		manifestConfig := dsp.Spec.DevFlags.Manifests[0]
+		if err := odhdeploy.DownloadManifests(ctx, componentsv1.DataSciencePipelinesComponentName, manifestConfig); err != nil {
+			return err
+		}
+
+		if manifestConfig.SourcePath != "" {
+			rr.Manifests[0].Path = odhdeploy.DefaultManifestPath
+			rr.Manifests[0].ContextDir = componentsv1.DataSciencePipelinesComponentName
+			rr.Manifests[0].SourcePath = manifestConfig.SourcePath
+		}
+	}
+
+	return nil
+}
+
+func UnmanagedArgoWorkFlowExists(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
+	// Check preconditions if this is an upgrade
+	if rr.Instance.GetStatus().Phase != status.PhaseReady {
```

Comment:
not sure if this hold true anymore, the phase can be empty also on the first reconcile, when the DSP CR get created. If this is something related to upgrades, then I guess it should be part of the [upgrade process](https://github.com/opendatahub-io/opendatahub-operator/blob/feature-operator-refactor/pkg/upgrade/upgrade.go).

but I may be wrong, so I let @VaishnaviHire and @zdtsw to chime in.

Reply:
Okay thank you I will wait to hear back from then on what is needed there

Reply:
so this function is introduced when we set ArgoWF as one of the backend for pipeline.
the case we want to cover is: if the ArgoWF version shipped from ODH does not match the one already being installed in cluster for other usage pre-ODH, it can cause conflict later.
so the workaround by that time is:
if user already use ArgoWF in cluster, either we cannot have ODH pipeline enabled, or we should ask them to removed these resource then enable ODH pipeline.

I do not know what is the plan for this in long term, maybe @HumairAK or @anishasthana has insight from pipeline side?

Reply:
> the case we want to cover is: if the ArgoWF version shipped from ODH does not match the one already being installed in cluster for other usage pre-ODH, it can cause conflict later.

if I read this right, then I guess this holds true at any point in time, not only in case of upgarde as i.e. a user can always change the CRD outside of the control of the operator